airbyte-workers/src/test-integration/java/io/airbyte/workers/process/KubePodProcessIntegrationTest.java

```diff
@@ -184,16 +187,20 @@ public void testPortsReintroducedIntoPoolOnlyOnce() throws Exception {
     // run a background process to continuously consume available ports
     final var portsTaken = new ArrayList<Integer>();
     final var executor = Executors.newSingleThreadExecutor();
+    final var shouldContinueTakingPorts = new AtomicBoolean(true);
+    final var doneTakingPorts = new CountDownLatch(1);
 
     executor.submit(() -> {
-      try {
-        while (true) {
-          portsTaken.add(KubePortManagerSingleton.getInstance().take());
+      while (shouldContinueTakingPorts.get()) {
+        final var portTaken = KubePortManagerSingleton.getInstance().takeImmediately();
+
+        if (portTaken != null) {
+          LOGGER.info("portTaken = " + portTaken);
+          portsTaken.add(portTaken);
         }
-      } catch (final InterruptedException e) {
-        e.printStackTrace();
-        throw new RuntimeException(e);
       }
+
+      doneTakingPorts.countDown();
```

Comment:
how does the latch help this?

Reply:
We wait on this `assertTrue(doneTakingPorts.await(5, TimeUnit.SECONDS));` 

What this allows us to do is handle the case where it was taking an additional element in the thread before draining. Now it actually waits for the thread to complete using this mechanism so we guarantee the thread isn't touching the `portsTaken` list.

Reply:
Got it. Pity. I thought this would help with the socket hanging issue we are seeing.

Reply:
Sadly nope.